File: leetcode/BackTracking/47. 全排列 II(有重复).py
# coding=utf-8
"""
给定一个可包含重复数字的序列，返回所有不重复的全排列。

示例:

输入: [1,1,2]
输出:
[
  [1,1,2],
  [1,2,1],
  [2,1,1]
]

"""
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        if not nums:
            return []

        self.res = []
        nums.sort()

        # 这属于一个比较通用的解法，因为在全排序2会碰到有相同元素的情况
        self.visited = [False] * len(nums)
        self.helper(nums, [])

        return self.res

    def helper(self, nums, temp):
        if len(temp) > len(nums):
            return

        if len(temp) == len(nums):
            self.res.append(temp[:])

        for i in range(len(nums)):
            # 假如说当前这个元素已经用过了，我们跳过
            if self.visited[i] == True:
                continue

            # i>0,保证了我们可以用到nums的第一个元素
            # self.visited[i-1] == True and nums[i-1] == nums[i]保证了我们用的 nums[i]都是和之前没重复的
            if i > 0 and self.visited[i - 1] == True and nums[i - 1] == nums[i]:
                continue

            self.visited[i] = True
            self.helper(nums, temp + [nums[i]])
            self.visited[i] = False

        return

"""
相比于permutation, 唯一的区别就是多了一个去重的处理
1. nums = sorted(nums) , 排序好了才可以去重
2. if i > 0 and nums[i] == nums[i - 1] and self.visited[i - 1]
   采用了全局去重，i > 0, 于此同时，我们要注意 self.visited[i-1] == False
   因为去除 i的重，要保证i-1已经在 答案列表里面了 

"""

File: leetcode/BackTracking/47. 全排列 II(有重复).go
package BackTracking

import "sort"

func permuteUnique(nums []int) [][]int {

	var res [][]int
	var temp []int
	var dfs func(cur int, temp []int)
	// 利用这个Visited来做回溯
	visited := make(map[int]bool)
	sort.Ints(nums)

	dfs = func(cur int, temp []int) {
		if len(temp) > len(nums) {
			return
		}

		if len(temp) == len(nums) {
			res = append(res, append([]int(nil), temp...))
		}

		for i:= 0; i < len(nums); i++ {
			if visited[i] == true {
				continue
			}
			// TODO 为什么这里是i > 0而不是 和40一样，是i > cur呢？
			// 因为全排序要求从全局来看每个元素都要遍历
			// TODO visited[i-1] == true 为什么这样写呢
			// 因为我们只要求在本层循环中，不使用上层循环使用的元素
			if i > 0 && nums[i-1] == nums[i] && visited[i-1] == true{
				continue
			}
			temp = append(temp, nums[i])
			visited[i] = true
			dfs(i, temp)
			temp = temp[:len(temp)-1]
			visited[i] = false

		}
	}

	dfs(0, temp)
	return res
}
